### DISCUSSION TITLE
win32console.STD_{{IN,OUT}PUT,ERROR}_HANDLE should be non-negative

### DISCUSSION THREAD

**Discussion by wkschwartz** [mhammond/pywin32#2710] [Q&A] [answered]
# Source of confusion

The Python standard library and pywin32 both provide the STD_{{IN,OUT}PUT,ERROR}_HANDLE constants and they're not equal (the latter is the two's complement negative of the former). In Python 3.8, after doing `pip install pywin32`:

```python-console
>>> import subprocess
>>> from win32 import win32console
>>> subprocess.STD_OUTPUT_HANDLE
4294967285
>>> win32console.STD_OUTPUT_HANDLE
-11
>>> subprocess.STD_OUTPUT_HANDLE == win32console.STD_OUTPUT_HANDLE + 2**32
True
```

# What the docs say
According to [`GetStdHandle`](https://docs.microsoft.com/en-us/windows/console/getstdhandle#parameters), the `GET_*_HANDLE` constants have the following types and values

> **STD_INPUT_HANDLE** (DWORD) -10
> **STD_OUTPUT_HANDLE** (DWORD) -11
> **STD_ERROR_HANDLE** (DWORD) -12

(According to [Windows Data Types](https://docs.microsoft.com/en-us/windows/win32/winprog/windows-data-types), a `DWORD` is an `unsigned long` and 32-bits wide.) I was initially confused when I read the above and thought Microsoft was claiming that `STD_INPUT_HANDLE` was both a `DWORD` and -10. But https://github.com/MicrosoftDocs/Console-Docs/issues/177#issuecomment-728360313 confirms that `(DWORD)-10` is to be read a C cast, so the positive value in `subprocess` is correct. 

# Implementation notes
pywin32 defines these constants here:

https://github.com/mhammond/pywin32/blob/85e7b248c39a51a2b495afc028b13ecee05f91c3/win32/src/win32consolemodule.cpp#L2176-L2179

To store `(DWORD)-10` rather than -10, you wouldn't be able to use [`PyModule_AddIntConstant`](https://docs.python.org/3/c-api/module.html#c.PyModule_AddIntConstant) anymore. Probably the easiest thing to do is re-export the values in `subprocess`. Or you could do what CPython [does in `Modules/_winapi.c`](https://github.com/python/cpython/blob/9cc9e277254023c0ca08e1a9e379fd89475ca9c2/Modules/_winapi.c#L2009-L2011):

```c
#define F_DWORD "k"
// ...snip...
#define WINAPI_CONSTANT(fmt, con) \
    do { \
        PyObject *value = Py_BuildValue(fmt, con); \
        if (value == NULL) { \
            return -1; \
        } \
        if (PyDict_SetItemString(d, #con, value) < 0) { \
            Py_DECREF(value); \
            return -1; \
        } \
        Py_DECREF(value); \
    } while (0)
// ... snip ...
static int winapi_exec(PyObject *m)
{
    WinApiState *st = winapi_get_state(m);


    st->overlapped_type = (PyTypeObject *)PyType_FromModuleAndSpec(m, &winapi_overlapped_type_spec, NULL);
    if (st->overlapped_type == NULL) {
        return -1;
    }


    if (PyModule_AddType(m, st->overlapped_type) < 0) {
        return -1;
    }


    PyObject *d = PyModule_GetDict(m);
    // ...snip...
    WINAPI_CONSTANT(F_DWORD, STD_INPUT_HANDLE);
    WINAPI_CONSTANT(F_DWORD, STD_OUTPUT_HANDLE);
    WINAPI_CONSTANT(F_DWORD, STD_ERROR_HANDLE);
```

**Comment by kxrob** [comment]
Did that cause a problem?
The constants seem to work with their modules - and crossover.
`unsigned long` / DWORD has no equivalent in the standard Python types (but ctypes).
Yet  `(unsigned long) -11` seems to be even better represented by -11 in a generic sense: Assume sb hypothetically compiles stuff directly for a machine with 64bit unsigned long (Linux) and the constant somehow walks via pickle or so, then -11 would still be cast correctly, while 4294967285 fails.

**Comment by wkschwartz** [comment]
I'm unclear why 4294967285 wouldn't round-trip through Pickle correctly to `(DWORD)-11`, which equals `4294967285UL`.

Anyway this isn't a big deal. There was no code-correctness problem. I just found the conflict with the standard library confusing.

**Comment by kxrob** [comment, marked as answer]
> I'm unclear why 4294967285 wouldn't round-trip through Pickle correctly to `(DWORD)-11`, which equals `4294967285UL`.

The `(unsigned long) -11` compiled on such 64bit integer system would be 18446744073709551605UL.
A literally remembered Python value 4294967285 would not match in GetStdHandle etc., while -11 still does.
